### DISCUSSION TITLE
Sceptre http handler causes extra characters in template

### DISCUSSION THREAD

**Discussion by zaro0508** [cloudreach/sceptre#1238] [Q&A] [answered]
We get extra `&#39` and `&#34` characters in our template only when it's downloaded using the http sceptre template handler

environment
* sceptre ver 2.7.1
* python ver 3.7
* MAC/OS Monterey ver 12.0.1

Steps to reproduce:

Create the following sceptre template config. 
```
template:
  type: "http"
  url: "https://raw.githubusercontent.com/Sage-Bionetworks/aws-infra/v0.3.7/templates/ssm-parameters.j2"
stack_name: test-stack
sceptre_user_data:
  Prefix: /nba/
  Parameters:
    - Name: TeamToRoleArnMap
      Value: >-
        '[
          {"division":"east","team":"blazers"},
          {"mascot":"beavers","home":"oregon"}
        ]'
```

generate cloudformation template..
```
➜  sceptre generate develop/test-stack.yaml
---

Description: "Setup parameters in the SSM parameter store"
AWSTemplateFormatVersion: "2010-09-09"
Resources:
  TeamToRoleArnMap:
      Type: AWS::SSM::Parameter
      Properties:
        Name: /nba/TeamToRoleArnMap
        Value: &#39;[
  {&#34;division&#34;:&#34;east&#34;,&#34;team&#34;:&#34;blazers&#34;},
  {&#34;mascot&#34;:&#34;beavers&#34;,&#34;home&#34;:&#34;oregon&#34;}
]&#39;
        Type: 'String'
```

Actual behaviour:
Notice the  `&#39` and `&#34` characters in the `Value` property.

Expected behavior:
Now download the jinja template `https://raw.githubusercontent.com/Sage-Bionetworks/aws-infra/v0.3.7/templates/ssm-parameters.j2` and save it to the local `template` folder.  Generate the cloudformation using the local jinja file and it will generate without those extra characters
```
template:
  path: "ssm-parameters.j2"
stack_name: test-stack
sceptre_user_data:
  Prefix: /nba/
  Parameters:
    - Name: TeamToRoleArnMap
      Value: >-
        '[
          {"division":"east","team":"blazers"},
          {"mascot":"beavers","home":"oregon"}
        ]'
```

```
➜  sceptre generate develop/test-stack.yaml
---
Description: "Setup parameters in the SSM parameter store"
AWSTemplateFormatVersion: "2010-09-09"
Resources:
  TeamToRoleArnMap:
      Type: AWS::SSM::Parameter
      Properties:
        Name: /nba/TeamToRoleArnMap
        Value: '[
  {"division":"east","team":"blazers"},
  {"mascot":"beavers","home":"oregon"}
]'
        Type: 'String'
```

**Comment by jfalkenstein** [comment]
That's weird. Have you figured out what is causing this?

**Comment by zaro0508** [comment]
Nope, haven't had time.  My current workaround is to download the file with the sceptre `before_launch` hook.

**Comment by jfalkenstein** [comment]
My first impulse would be to look at encoding issues here. Perhaps we're not properly decoding the response or it's not encoded in utf-8 but instead something different.

**Comment by zaro0508** [comment, marked as answer]
I have finally figured this out... 

The Sceptre jinja environment's [autoescape setting is enabled by default](https://github.com/Sceptre/sceptre/blob/master/sceptre/template_handlers/helper.py#L130-L134) therefore jinja is auto escaping the characters when rendering the template.

There are two ways to get around this problem:
* Use the jinja [safe filter](https://jinja.palletsprojects.com/en/3.1.x/templates/#jinja-filters.safe) in the jinja template to mark the variable as safe (i.e. `{{ VAR|safe }}`)
* Disable the jinja environment's autoescape setting in the [Sceptre config.yaml file](https://docs.sceptre-project.org/latest/docs/stack_group_config.html#j2-environment):
```
j2_environment:
   autoescape: False
```

**Reply by iainelder** [reply]
Ran into this now. Thanks @zaro0508 for helping to understand the behavior and how to work around it.

I made a Sceptre project with different combinations of template type and autoescape setting. The GitHub workflow generates the YAML output for each one.

https://github.com/iainelder/sceptre-http-jinja2-bug/actions/runs/6593703005/job/17916559230

```
Case: file-autoescape-false/test.yaml
---
Unsafe: "Hello, world!"
Safe: "Hello, world!"

Case: file-autoescape-true/test.yaml
---
Unsafe: &#34;Hello, world!&#34;
Safe: "Hello, world!"

Case: file/test.yaml
---
Unsafe: "Hello, world!"
Safe: "Hello, world!"

Case: http-autoescape-false/test.yaml
---
Unsafe: "Hello, world!"
Safe: "Hello, world!"

Case: http-autoescape-true/test.yaml
---
Unsafe: &#34;Hello, world!&#34;
Safe: "Hello, world!"

Case: http-github/test.yaml
---
Unsafe: &#34;Hello, world!&#34;
Safe: "Hello, world!"

Case: http-local/test.yaml
---
Unsafe: &#34;Hello, world!&#34;
Safe: "Hello, world!"
```
